src/grammar/cursor.rs

```diff
@@ -0,0 +1,487 @@
+use crate::{safe_loop, Parser};
+use source_gen::{lexer::TokenKind, syntax::SyntaxKind, T};
+
+use super::{parse_datatype, parse_expr, parse_ident, parse_query};
+
+/// Railroad diagram 🚆 https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/explicit-cursor-declaration-and-definition.html
+pub fn parse_cursor(p: &mut Parser) {
+    p.start(SyntaxKind::CursorStmt);
+    p.expect(T![cursor]);
+    parse_ident(p, 1..1);
+    if p.at(T!["("]) {
+        parse_cursor_param_declarations(p);
+    }
+    if p.eat(T![return]) {
+        parse_rowtype_clause(p);
+    }
+    if p.eat(T![is]) {
+        parse_query(p, false);
+    }
+    p.eat(T![;]);
+    p.finish();
+}
+
+fn parse_cursor_param_declarations(p: &mut Parser) {
+    p.start(SyntaxKind::CursorParameterDeclarations);
+    p.expect(T!["("]);
+    safe_loop!(p, {
+        parse_cursor_param_declaration(p);
+        if !p.eat(T![,]) {
+            break;
+        }
+    });
+    p.expect(T![")"]);
+    p.finish();
+}
+
+fn parse_cursor_param_declaration(p: &mut Parser) {
+    p.start(SyntaxKind::CursorParameterDeclaration);
+    parse_ident(p, 1..1);
+    p.eat(T![in]);
+    parse_datatype(p);
+    if [T![:=], T![default]].contains(&p.current()) {
+        p.eat(T![:=]);
+        p.eat(T![default]);
+        parse_expr(p);
+    }
+    p.finish();
+}
+
+fn parse_rowtype_clause(p: &mut Parser) {
+    p.start(SyntaxKind::RowtypeClause);
+    parse_ident(p, 1..2);
+    if p.at(T![%]) && p.nth(1) == Some(T![rowtype]) {
+        p.expect(T![%]);
+        p.expect(T![rowtype]);
```

Comment:
We're not covering the `%TYPE` here
![image](https://github.com/user-attachments/assets/848f4e05-2993-4bbf-90fc-95adb26c8605)

Reply:
is type the literal TYPE here?

Reply:
Yep, should work fine